resources/bundles/org.eclipse.core.filesystem/src/org/eclipse/core/internal/filesystem/InternalFileSystemCore.java

```diff
@@ -72,9 +81,15 @@ public IFileSystem getFileSystem(String scheme) throws CoreException {
 			IConfigurationElement element = (IConfigurationElement) result;
 			FileSystem fs = (FileSystem) element.createExecutableExtension("run"); //$NON-NLS-1$
 			fs.initialize(scheme);
-			//store the file system instance so we don't have to keep recreating it
-			registry.put(scheme, fs);
-			return fs;
+			synchronized (this) {
```

Comment:
You should move the part from the config element below the part that checks if there is already an instance

Reply:
> You should move the part from the config element below the part that checks if there is already an instance

No, we don't know if that code takes any locks (which is allowed) and so we might run into deadlock running that code with another lock aquired.

Reply:
But now you create (and init) a FS on each call but probably never use it...

Reply:
> But now you create (and init) a FS on each call but probably never use it...

Not on each call - only for these threads that might init this in parallel. Once initialized, the value from map is used. Extension points allow multiple creation of contributions, and changing existing implementation to use a lock to avoid that is again breaking client code.

Reply:
the returned map is unsyncronized, so it is not safe to modify it while other threads are already reading it (what could happen here). So then the first read must be syncronized as well, or a concurrent map must be used with computeIfAbsent if you don't like to lock ...

Reply:
Correct, I tend to forget that, see https://github.com/eclipse-platform/eclipse.platform/pull/1144